tutorials/haskell/mustache-templates/tutorial.md

```diff
@@ -0,0 +1,404 @@
+---
+
+[Mustache templates](https://mustache.github.io/) are popular logic-less
+templates that can be used in web development as well as in any other domain
+where interpolation of data into some sort of template is needed.
+“Logic-less templates” means that the templates themselves do not have any
+logic such as conditionals (although it's possible to choose how to
+represent empty collections for example) and thus it's much harder to shoot
+yourself in the foot when concerns of data processing and representation are
+separated.
+
+Libraries supporting working with Mustache templates are available for many
+languages, including Haskell (there are several of them, actually). In this
+tutorial we will be using the
+[`stache`](https://hackage.haskell.org/package/stache) package developed by
+Stack Builders and show why this package is clearer and easy to use.
+
+The tutorial should be comprehensible for Haskell beginners and does not
+assume knowledge of any advanced Haskell-concepts.
+
+## Available libraries
+
+Let's see which libraries are available to work with Mustache templates:
+
+* [`hastache`](https://hackage.haskell.org/package/hastache) is the oldest
+  (first released in 2011) and now deprecated library in this list. Those
+  who used the library will probably agree that it has not the clearest API
+  where a lot of wrapping is necessary and dictionaries are represented as
+  wrapped functions. Generics can be used to alleviate the situation, using
+  yet another wrapper called `mkGenericContext`.
+
+* [`mustache`](https://hackage.haskell.org/package/mustache) is the official
+  successor of `hastache`, but it again makes simple things complex using
+  Aeson's `Value` (good) and at the same time introducing its own `Value`
+  type (with conflicting names of constructors and naturally not so numerous
+  instances). Source code is filled with Unicode symbols (using packages
+  like `base-unicode-symbols`, etc.), so it's not easy to edit it if you
+  want to contribute to the package.
+
+* [`stache`](https://hackage.haskell.org/package/stache) is an alternative
+  implementation that conforms to official Mustache spec and passes the
+  official test suite. Its API consists of only 4 functions (3 to compile
+  templates and one to render them using any instance of `ToJSON` as source
+  of data to interpolate), plus Template Haskell helpers to compile/validate
+  templates at compile time. Parser is written with Megaparsec 5. The
+  package uses `Data.Text.Lazy.Builder` under the hood and produces lazy
+  `Text`.
+
+The main motivation for developing `stache` was the desire to expose more
+minimal API and use Aeson's instances directly for value interpolation, as
+well as the desire to use Megaparsec instead of Parsec for parsing.
+Initially we wanted to contribute to the existing `mustache` library, but
+then realized that changes we want to implement are too cardinal and we are
+better off writing our own package.
```

Comment:
@mrkkrp "but then realized that the changes we wanted to implement were too cardinal and we were better off writing our own package."